### PR TITLE
Add GAN-based TTS Vocoder

### PR DESCRIPTION
**Pull request by blisc**
- Adds the fullband MelGAN model as described in https://arxiv.org/abs/2005.05106
- Remove empty *_step() from ModelPT
- Split UniGlowLoss into glow portion and STFTLoss portion
- Simply stft_exact_pad option in FilterBankFeatures

ToDos
- [x] ~Enable FP16~. No currently a priority.
- [x] Add NeMo Loss subclass
- [x] Update Docs
- [x] ~Update TTS tutorials~, can't until I upload the .nemo files

### CODE REVIEW

File: nemo/collections/tts/models/melgan.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        return torch.utils.data.DataLoader(dataset, collate_fn=dataset.collate_fn, **cfg.dataloader_params)
+
+    def setup_training_data(self, cfg):
+        self._train_dl = self.__setup_dataloader_from_config(cfg)
+
+    def setup_validation_data(self, cfg):
+        self._validation_dl = self.__setup_dataloader_from_config(cfg, shuffle_should_be=False, name="validation")
+
+    def training_epoch_end(self, outputs):
+        # Do manual logginging of learning rate and epoch
```
Comment (stasbel) [inline nemo/collections/tts/models/melgan.py:300]: Why manual?

**Review by stasbel — APPROVED** [review]

File: nemo/collections/tts/models/melgan.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+from nemo.collections.tts.models.base import Vocoder
+from nemo.core.classes.common import typecheck
+from nemo.core.neural_types.elements import AudioSignal, MelSpectrogramType
+from nemo.core.neural_types.neural_type import NeuralType
+from nemo.core.optim.lr_scheduler import CosineAnnealing
+from nemo.utils import logging
+
+
+class MelGanModel(Vocoder):
+    def __init__(self, cfg: DictConfig, trainer: 'Trainer' = None):
```
[the remark this answers is not in the document]
Comment (blisc) [reply nemo/collections/tts/models/melgan.py:30]: Yes, it's the preferred way of doing typehints without importing the class.
See: https://www.python.org/dev/peps/pep-0484/#forward-references

File: nemo/collections/tts/models/melgan.py (hunk 1 above)
Comment (blisc) [reply nemo/collections/tts/models/melgan.py:300]: I could never get lightning's self.log to work properly, so I reverted to tensorboard's API
